test/testexprengine.cpp

```diff
@@ -442,6 +453,109 @@ class TestExprEngine : public TestFixture {
                       expr("void foo(const int *x) { if (f1() && *x > 12) dostuff(*x == 5); }", "=="));
     }
 
+    void ifAlwaysTrue1() {
+        const char code[] = "int foo() {\n"
+                            "  int a = 42;\n"
+                            "  if (1)\n"
+                            "    a = 0;\n"
+                            "  return a == 0;\n"
+                            "}";
+        const char expected[] = "(distinct 1 0)\n"
+                                "(= 0 0)\n"
+                                "z3::sat\n";
+        ASSERT_EQUALS(expected, expr(code, "=="));
+    }
+
+    void ifAlwaysTrue2() {
+        const char code[] = "int foo() {\n"
+                            "  if (1.0)\n"
+                            "    a = 0;\n"
+                            "  return a == 0;\n"
+                            "}";
+        const char expected[] = "(distinct 1 0)\n"
+                                "(= 0 0)\n"
+                                "z3::sat\n";
+        TODO_ASSERT_EQUALS(expected, "", expr(code, "=="));
+    }
+
+    void ifAlwaysTrue3() {
+        const char code[] = "int foo() {\n"
+                            "  int a = 42;\n"
+                            "  if (\"foo\")\n"
+                            "    a = 0;\n"
+                            "  return a == 0;\n"
+                            "}";
+        const char expected[] = "(distinct 1 0)\n"
+                                "(= 0 0)\n"
+                                "z3::sat\n";
+        TODO_ASSERT_EQUALS(expected, "", expr(code, "=="));
+    }
+
+    void ifAlwaysTrue4() { // Known local variable
+        const char code[] = "int foo() {\n"
+                            "  float a = 42.0;\n"
+                            "  if (a == 42.0)\n"
+                            "    a = 0.3;\n"
+                            "  return a == 0.3;\n"
+                            "}";
+        const char expected[] = "(= |42.0| |42.0|)\n"
+                                "z3::sat\n"
+                                "(= |42.0| |42.0|)\n"
+                                "(= |0.3| |0.3|)\n"
+                                "z3::sat\n";
+        ASSERT_EQUALS(expected, expr(code, "=="));
+    }
+
+    void ifAlwaysFalse1() {
+        const char code[] = "int foo() {\n"
+                            "  int a = 42;\n"
+                            "  if (0)\n"
+                            "    a = 0;\n"
+                            "  return a == 0;\n"
+                            "}";
+        const char expected[] = "(= 0 0)\n"
+                                "(= 42 0)\n"
+                                "z3::unsat\n";
+        ASSERT_EQUALS(expected, expr(code, "=="));
+    }
+
+    void ifAlwaysFalse2() {
+        const char code[] = "int foo() {\n"
+                            "  if (0.0)\n"
+                            "    a = 0;\n"
+                            "  return a == 0;\n"
```

Comment:
aha.. please declare `a`. It will not get a ExprEngine value if it does not have a varid.

Reply:
Ah. Thanks.
This allowed me to fix that test, but now I get segfault in another test. I'm working on it.

Reply:
There was a bug in z3: https://github.com/Z3Prover/z3/issues/4905.